.circleci/config.yml

```diff
@@ -1,42 +1,95 @@
+      - MKL_NUM_THREADS: 2
+      - OPENBLAS_NUM_THREADS: 2
+      - CONDA_ENV_NAME: testenv
+      - LOCK_FILE: build_tools/circle/doc_min_dependencies_linux-64_conda.lock
+      # Sphinx race condition in doc-min-dependencies is causing job to stall
+      # Here we run the job serially
+      - SPHINX_NUMJOBS: 1
     steps:
       - checkout
-      - run: bash build_tools/circle/download_documentation.sh
+      - run: ./build_tools/circle/checkout_merge_commit.sh
+      - restore_cache:
+          key: v1-doc-min-deps-datasets-{{ .Branch }}
+      - restore_cache:
+          keys:
+            - doc-min-deps-ccache-{{ .Branch }}
+            - doc-min-deps-ccache
+      - run: ./build_tools/circle/build_doc.sh
+      - save_cache:
+          key: doc-min-deps-ccache-{{ .Branch }}-{{ .BuildNum }}
+          paths:
+            - ~/.ccache
+            - ~/.cache/pip
+      - save_cache:
+          key: v1-doc-min-deps-datasets-{{ .Branch }}
+          paths:
+            - ~/scikit_learn_data
       - store_artifacts:
           path: doc/_build/html/stable
           destination: doc
+      - store_artifacts:
+          path: ~/log.txt
+          destination: log.txt
 
   doc:
     docker:
       - image: cimg/python:3.8.12
     environment:
-      - GITHUB_ARTIFACT_URL: << pipeline.parameters.GITHUB_RUN_URL >>/doc.zip
+      - MKL_NUM_THREADS: 2
+      - OPENBLAS_NUM_THREADS: 2
+      - CONDA_ENV_NAME: testenv
+      - LOCK_FILE: build_tools/circle/doc_linux-64_conda.lock
     steps:
       - checkout
-      - run: bash build_tools/circle/download_documentation.sh
+      - run: ./build_tools/circle/checkout_merge_commit.sh
+      - restore_cache:
+          key: v1-doc-datasets-{{ .Branch }}
+      - restore_cache:
+          keys:
+            - doc-ccache-{{ .Branch }}
+            - doc-ccache
+      - run: ./build_tools/circle/build_doc.sh
+      - save_cache:
+          key: doc-ccache-{{ .Branch }}-{{ .BuildNum }}
+          paths:
+            - ~/.ccache
+            - ~/.cache/pip
+      - save_cache:
+          key: v1-doc-datasets-{{ .Branch }}
+          paths:
+            - ~/scikit_learn_data
       - store_artifacts:
           path: doc/_build/html/stable
           destination: doc
-      # Persists the generated documentation, so that it
-      # can be attached and deployed in the "deploy" job
+      - store_artifacts:
+          path: ~/log.txt
+          destination: log.txt
+      # Persists generated documentation so that it can be attached and deployed
+      # in the 'deploy' step.
       - persist_to_workspace:
           root: doc/_build/html
           paths: .
 
+  lint:
+    docker:
+      - image: cimg/python:3.8.12
+    steps:
+      - checkout
+      - run: ./build_tools/circle/checkout_merge_commit.sh
+      - run:
+          name: dependencies
+          command: pip install flake8
+      - run:
+          name: linting
+          command: ./build_tools/azure/linting.sh
```

Comment:
Actually it is what you did here.

Reply:
Maybe we could just remove the `lint` job from the circle ci config since we already done on Azure. No need to do it twice, no?

That would simplify the CI status summary and make it simpler for first time contributors not to have to deal with two slightly different linter logs.

I think it's fine building the doc even if flake8 does not pass.

Reply:
I like linting on CircleCI, because it helps saves credits on CircleCI.

I do not want to block this PR, so I am okay with merging first and thinking about linting as a follow up.